### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by wooorm**
Okay! Here goes. Most of this is just syntactic rewrite to how I like my code, which is just to ensure I don’t have to change small stuff in the future (e.g., setting variable declarations at the top of a scope, strict equality checks), but also because it helps me understand the code.

The most important change has to do with regexes: I’m trying to minimise those and rewrote the `escape` method to use an algorithm instead. I also tried my hand at `unescapeShortcutLinkRefer` (but didn’t get it working yet, and time is running out today). That’s a funky function, for example, I’m not sure what the `+` does in the regex for `PUNCTUATION`? Does it ensure the backslash is included?

Anyway, it doesn’t work yet, but I was hoping with your understanding of the regex and the function could get it to work? If not, just cherry pick this commit and ignore that function!

### CODE REVIEW

File: lib/stringify.js [hunk 1]
```diff
@@ -1520,12 +1619,61 @@ function label(node) {
...
-        return canonicalPunctuation[index++];
-    });
+    while (++index < length) {
+        character = value.charAt(index);
+
+        if (identifier.charAt(index) === BACKSLASH) {
+            next = identifier.charAt(index + 1);
+
+            if (
+                next === DASH ||
```
Comment (wooorm) [inline lib/stringify.js:1635]: Is the `_` missing here? I didn’t see it in the `PUNCTUATION` regex!

File: lib/stringify.js [hunk 2]
```diff
@@ -308,50 +317,134 @@ function escapeFactory(options) {
...
-                    --i >= 0 &&
-                    parent.children[i].type == 'text'
-                )
+            /*
+             * Ensure non-auto-links are not seen links.
+             * This pattern needs to check the preceding
+             * nodes too.
+             */
+
+            if (prev && prev.type == 'text' && value.charAt(0) == COLON) {
```
Comment (wooorm) [inline lib/stringify.js:429]: Shouldn’t this be `commonmark` only (or even anything non-GFM)? Just like the one-node method above?

File: lib/stringify.js [hunk 3]
```diff
@@ -1520,12 +1619,61 @@ function label(node) {
...
+            ) {
+                character = value.charAt(index + 1);
+                index++;
+            }
+        }
+
+        result += character;
+    }
+
+    console.log('result: ', [result, value, identifier]);
```
Comment (eush77) [inline lib/stringify.js:1675]: Left from debugging?

File: lib/stringify.js (hunk 1 above)
Comment (eush77) [inline lib/stringify.js:1635]: `_` is missing indeed. Nice catch!

File: lib/stringify.js (hunk 3 above)
Comment (wooorm) [inline lib/stringify.js:1675]: Yeah ;) note, as mentioned in the PR, this fn doesn't work yet!

File: lib/stringify.js [hunk 4]
```diff
@@ -308,50 +317,134 @@ function escapeFactory(options) {
...
+                    )
+                )
+            ) {
+                queue += BACKSLASH;
+            } else if (character === LINE) {
+                afterNewLine = true;
+            } else if (afterNewLine) {
+                if (
+                    character === ANGLE_BRACKET_CLOSE ||
+                    character === DASH
```
Comment (wooorm) [inline lib/stringify.js:366]: @eush77 Should `+` be handled here?

**Comment by eush77** [comment]
> I’m not sure what the `+` does in the regex for `PUNCTUATION`? Does it ensure the backslash is included?

`+` can be escaped in Commonmark mode, so `unescapeShortcutLinkReference("a\\+b", "a+b")` should return `"a+b"`. Conversely, `unescapeShortcutLinkReference("a+b", "a\\+b")` should return `"a\\+b"`, so we can't restrict `PUNCTUATION` regex to only those character that we escape during stringification.

Originally I thought of replacing groups of consecutive non-letters, but I soon realized that matching letters is tricky due to lack of unicode regexes and unicode-aware character classes in JavaScript (at least in ES5). Then I came up with `PUNCTUATION` regex which is generated from `require('./lib/expressions').commonmark.escape` and matches character groups that might deviate from `identifier` in a way that must be fixed after stringification.

File: lib/stringify.js (hunk 4 above)
Comment (eush77) [inline lib/stringify.js:366]: Yes! I forgot `+` can be a bullet character.

File: lib/stringify.js (hunk 2 above)
Comment (eush77) [inline lib/stringify.js:429]: Yeah, probably GFM-only.

**Comment by wooorm** [comment]
Regarding the `+` in `PUNCTUATION`, I think I was a bit vague there. I didn’t mean the `+` inside the capturing group, rather, the `+` outside it ([just before the `/g`](https://github.com/eush77/mdast/blob/7b92078711a17f21264fe98381ee1c5539bdc455/lib/stringify.js#L70)). Wouldn’t that catch more characters (for example, `'--:;'`) than needed?

I think I’m understanding what happens in the `unescape` function: It takes all characters from `value`, except for backslashes, which come (or not) from `identifier`?

**Comment by eush77** [comment]
No, that is actually the whole point of this function: it replaces groups of consecutive punctuation characters from `value` with corresponding groups from `identifier`. This way it combines the original escaping from `identifier` (backslashes must be captured for this) with whitespace and uppercase/lowercase distinction from `value`.

**Comment by wooorm** [comment]
@eush77 Ah I understand. I hope it’s possible to do that with an algorithm instead of a regex, I hope you can investigate that. If not, as stated above, I’m fine with leaving that for later and just cherry picking the rest of the commits :smile:

**Pull request closed without merging**